### PR TITLE
Webauthn contract

### PR DESCRIPTION
**Pull request by akonior**

### CODE REVIEW

File: src/app/account/webauthn_account_contract.ts [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+  constructor(
+    readonly challenge: Uint8Array,
+    readonly authenticator_data: Uint8Array,
+    readonly client_data_json: Uint8Array,
+    readonly signatureRaw: Uint8Array
+  ) {
+  }
+}
+
+export interface WebAuntnInterface {
```
Comment (marekkirejczyk) [inline src/app/account/webauthn_account_contract.ts:42]: Typo? `WebAuntnInterface` -> `WebAuthnInterface`

File: src/app/components/webauthn/login.tsx [hunk 2]
```diff
@@ -52,6 +52,10 @@ function convertASN1toRaw(signatureBuffer: ArrayBuffer) {
     const sStart = usignature[rEnd + 2] === 0 ? rEnd + 3 : rEnd + 2;
     const r = usignature.slice(rStart, rEnd);
     const s = usignature.slice(sStart);
+    if (s[0] > 0x80) {
+        // TODO if is greater than secp256r1n/2 we need to flip s := secp256r1n - s
+        throw new Error('BAD SIGNATURE');
+    }
```
Comment (marekkirejczyk) [inline src/app/components/webauthn/login.tsx:58]: Wow, what a coincidence, we just investigated why we have to flip , just few weeks ago.

File: src/app/components/webauthn/login.tsx [hunk 3]
```diff
@@ -24,12 +24,12 @@ async function webAuthnLogin(challenge: Uint8Array) {
...
+        await webAuthnLogin(new Uint8Array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]));
 
-    const challenge_str = `challenge = [${challenge}}]`;
-    const client_data_json_str = `client_data_json = [${client_data_json}]`;
-    const authenticator_data_str = `authenticator_data = [${authenticator_data}]`;
-    const signature_str = `signature = [${new Uint8Array(signatureRaw)}]`;
+    const challenge_str = `let challenge = [${challenge}];`;
+    const client_data_json_str = `let client_data_json = [${client_data_json}];`;
+    const authenticator_data_str = `let authenticator_data = [${authenticator_data}];`;
+    const signature_str = `let signature = [${new Uint8Array(signatureRaw)}];`;
```
Comment (marekkirejczyk) [inline src/app/components/webauthn/login.tsx:32]: What kind of dark magic is that?
Why not just `console.log(\`let challenge = [${challenge}];\`)`

File: src/app/components/webauthn/register.tsx [hunk 4]
```diff
@@ -43,10 +43,10 @@ async function webAuthnFetchPublicKey(challenge: ArrayBuffer): Promise<{ x: any,
 }
 
 const register = async () => {
-    const challenge = new Uint8Array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,]);
+    const challenge = new Uint8Array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]);
     const {x, y} = await webAuthnFetchPublicKey(challenge);
-    const pub_key_x_str = `pub_key_x = [${x}]`;
-    const pub_key_y_str = `pub_key_y = [${y}]`;
+    const pub_key_x_str = `let pub_key_x = [${x}];`;
+    const pub_key_y_str = `let pub_key_y = [${y}];`;
```
Comment (marekkirejczyk) [inline src/app/components/webauthn/register.tsx:49]: #darkmagic

File: src/contracts/webauthn_account_contract/src/public_key_note.nr [hunk 5]
```diff
@@ -13,19 +13,19 @@ use dep::aztec::{
...
-global PUBLIC_KEY_NOTE_LEN: Field = 3;
+global PUBLIC_KEY_NOTE_LEN: Field = 5;
 
 // Stores a public key composed of two fields
 // TODO: Do we need to include a nonce, in case we want to read/nullify/recreate with the same pubkey value?
 struct PublicKeyNote {
-    x: Field,
-    y: Field,
+    x: [u8; 32],
+    y: [u8; 32],
```
Comment (marekkirejczyk) [inline src/contracts/webauthn_account_contract/src/public_key_note.nr:22]: Some considerations for the future:
1) Can we make it a type?
2) Would it be more efficient if do [Field; 2], where x = [x1, x2], x1 = x / p, x2 = x % p?
Feels like a lot of constraints.

File: src/contracts/webauthn_account_contract/src/public_key_note.nr [hunk 6]
```diff
@@ -34,9 +34,30 @@ impl PublicKeyNote {
...
+        let mut y: Field = 0;
+        let mut mul: Field = 1;
+
+        for i in 1..32 {
+          let byte_x: Field = self.x[31 - i] as Field;
+          x = x + (byte_x * mul);
+          let byte_y: Field = self.y[31 - i] as Field;
+          y = y + (byte_y * mul);
+          mul *= 256;
+        }
```
Comment (marekkirejczyk) [inline src/contracts/webauthn_account_contract/src/public_key_note.nr:54]: Nice!

File: src/app/account/webauthn_account_contract.ts (hunk 1 above)
Comment (akonior) [reply src/app/account/webauthn_account_contract.ts]: Fixed

File: src/app/components/webauthn/login.tsx (hunk 2 above)
Comment (akonior) [reply src/app/components/webauthn/login.tsx:58]: Exactly.

**Review by marekkirejczyk — APPROVED** [review]

File: src/app/components/webauthn/login.tsx (hunk 3 above)
Comment (akonior) [reply src/app/components/webauthn/login.tsx:32]: A this point I wanted to manually check if WebAuth output can be properly checked on Noir contract side. So just for convenience I want to output proper Noir code in one console output for easier copy-paste.

**Pull request merged**